File: linkedlist/mergeNodesInBetweenZeroes.java
package linkedlist;

public class mergeNodesInBetweenZeroes{
    public static void main(String[] args) {
        System.out.println("Interactive Problem So Test The Solution in Leetcode Itself");
    }
}


class SolutionMNIBZ {

    // Submitted by @Jiganesh
    
    // TC : O(n)
    // SC : O(1)

    public ListNode mergeNodes(ListNode head) {
        
        
        ListNode current = head;
        int summation = 0;
        
        while(current != null){
            if (current.val ==0){
                current.val =summation;
                summation=0;
            }else{
                summation+=current.val;
                current.val=0;
            }
            current= current.next;
        }
        
        current = head;
        while (current != null && current.next !=null){
            if (current.next.val==0  && current.next.next !=null){
                current.next = current.next.next;
            }else{
                current= current.next;
            }
        }
        return head.next;
    }
}

File: linkedlist/mergeNodesInBetweenZeroes.py
# https://leetcode.com/problems/merge-nodes-in-between-zeros/

# Definition for singly-linked list.

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    
    # Submitted By @Jiganesh
    
    # Runtime: 4388 ms, faster than 100.00% of Python online submissions for Merge Nodes in Between Zeros.
    # Memory Usage: 131.5 MB, less than 100.00% of Python online submissions for Merge Nodes in Between Zeros.

    def mergeNodes(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """


        curr = head
        s = 0

        while curr:
            if curr.val ==0:
                curr.val = s
                s=0
            else:
                s+=curr.val
                curr.val=0
            curr=curr.next


        curr = head
        while curr and curr.next:
            if curr.next.val==0 and curr.next.next:
                curr.next = curr.next.next
            else:
                curr= curr.next

        return head.next

